### PR TITLE
Fix incorrect type definitions in useCallback

### PR DESCRIPTION
**Pull request by mununki**
This fixes the incorrect type definitions of `React.useCallback7, 8`.

### CODE REVIEW

**Comment by cristianoc** [comment]
@cknitt is there a better way than using all these versions?

**Comment by cknitt** [comment]
@mununki Ooops! Thanks for the fix!

Maybe something like

```rescript
external useCallback6: ('callback, ('a, 'b, 'c, 'd, 'e, 'f)) => 'callback = "useCallback"
```

would be nicer than

```rescript
external useCallback6: ('f, ('a, 'b, 'c, 'd, 'e, 'f0)) => 'f = "useCallback"
```

though?

**Comment by mununki** [comment]
> @cknitt is there a better way than using all these versions?

This reminds me of something: do we really need to use useCallback0-8 to make the number of dependent arrays type safe?

**Comment by cknitt** [comment]
> @cknitt is there a better way than using all these versions?

I was toying with the idea of introducing an abstract type for dependency arrays so that you would write

```rescript
React.useCallback(() => myCallback(a, b, c, d), React.deps4(a, b, c, d))
React.useEffect(() => myEffect(a), React.dep(a))
React.useMemo(() => myFunction(a, b), React.deps2(a, b))
```

Not sure if that would be an improvement though...

**Review by cknitt — APPROVED** [review]

**Comment by mununki** [comment]
> > @cknitt is there a better way than using all these versions?
> 
> I was toying with the idea of introducing an abstract type for dependency arrays so that you would write
> 
> ```rescript
> React.useCallback(() => myCallback(a, b, c, d), React.deps4(a, b, c, d))
> React.useEffect(() => myEffect(a), React.dep(a))
> React.useMemo(() => myFunction(a, b), React.deps2(a, b))
> ```
> 
> Not sure if that would be an improvement though...

Oh, I meant using just array:
```res
React.useCallback3(_ => f, [a,b,c])
```

**Comment by cristianoc** [comment]
Well if you have deps, then it can as well take a generic `'a` so it does not constrain what you pass.
What exactly is and is not allowed as dependency?

**Comment by cknitt** [comment]
> Oh, I meant using just array:
> 
> ```rescript
> React.useCallback(_ => f, [a,b,c])
> ```

This will only work if a, b, c have the same type.

**Comment by cristianoc** [comment]
The array forces you to inject everything into a single type, which complicates things.

**Comment by cknitt** [comment]
The bindings are trying to enforce that the dependency array is in fact an array.

If we just type the dependency array as `'deps` instead then you can pass what you want:

```rescript
React.useCallback(() => myCallback(a, b, c, d), (a, b, c, d))
React.useEffect(() => myEffect(a), [a])
React.useMemo(() => myFunction(a, b), (a, b))
```

**Comment by cknitt** [comment]
But would be less safe and allow people to pass

```rescript
React.useEffect(() => myEffect(a), a)
```

instead of

```rescript
React.useEffect(() => myEffect(a), [a])
```

for example

**Comment by cristianoc** [comment]
> The bindings are trying to enforce that the dependency array is in fact an array.
> 
> If we just type the dependency array as `'deps` instead then you can pass what you want:
> 
> ```rescript
> React.useCallback(() => myCallback(a, b, c, d), (a, b, c, d))
> React.useEffect(() => myEffect(a), [a])
> React.useMemo(() => myFunction(a, b), (a, b))
> ```

Is there any serious risk of passing something that is not allowed i.e. is not picked up, or is it pretty clear this is low risk in practice?

**Comment by cristianoc** [comment]
E.g. can you pass a record/object? I guess no.

**Comment by cometkim** [comment]
React team actually working on some compiler for it (codename: Forget)
https://react.dev/blog/2022/06/15/react-labs-what-we-have-been-working-on-june-2022#react-compiler

So.. we may end up needing PPX or something to build equivalents.

**Comment by cristianoc** [comment]
E.g. one possible mistake would be to pass a list, but I'm not sure we want to worry about that.
In future, we can always add a more precise type to the language.

**Comment by mununki** [comment]
Maybe heterogeneous array for deps.

**Comment by cknitt** [comment]
Ok, what happens if I pass for example a number instead of an array is that I get this warning:

> Warning: useMemo received a final argument that is not an array (instead, received `number`). When specified, the final argument must be an array.

**Comment by mununki** [comment]
> Currently, in addition to the useCallback1, 2, ... we have this binding
> 
> ```rescript
> @module("react")
> external useCallback: 'f => 'f = "useCallback"
> ```
> 
> which does not really make sense, as without a dependency array nothing is memoized.
> 
> So we could just change that one to
> 
> ```rescript
> @module("react")
> external useCallback: ('f, 'deps) => 'f = "useCallback"
> ```
> 
> so that we have a less safe, but more convenient option in addition to the useCallback1, 2, ...

How about removing the useCallbacks? It seems to me that useCallbacks0-7 are the ones that guarantee there are deps.

**Comment by mununki** [comment]
I've tried to define a heterogeneous list something like:
```res
type rec dep<'b, 'c> =
  | Nil: dep<'b, 'b>
  | Cons('a, dep<'b, 'c>): dep<'a, 'b>

let x = Cons(1, Cons("a", Cons(1.5, Nil)))
```
But it doesn't compile with `@unboxed`.

**Comment by mununki** [comment]
> Currently, in addition to the useCallback1, 2, ... we have this binding
> 
> ```rescript
> @module("react")
> external useCallback: 'f => 'f = "useCallback"
> ```
> 
> which does not really make sense, as without a dependency array nothing is memoized.
> 
> So we could just change that one to
> 
> ```rescript
> @module("react")
> external useCallback: ('f, 'deps) => 'f = "useCallback"
> ```
> 
> so that we have a less safe, but more convenient option in addition to the useCallback1, 2, ...

https://github.com/rescript-lang/rescript-react/pull/97/commits/f1422b42e3d5af220bf307d36cc2d24cfefe58ef

**Comment by mununki** [comment]
Same goes to `useEffect` and `useMemo`?

**Comment by cknitt** [comment]
> How about removing the useCallback? It seems to me that useCallbacks0-7 are the ones that guarantee there are deps.

> Same goes to `useEffect` and `useMemo`?

Same thing for all three, we have two options:

1. Remove the binding without a dependency array
2. Add a dependency array to that binding, typed as `'deps` for convenience.

I understood @cristianoc that he would like something like option 2 as we are not too worried about errors here:

> E.g. one possible mistake would be to pass a list, but I'm not sure we want to worry about that.
> In future, we can always add a more precise type to the language.

Usage would be like this:

```rescript
React.useCallback(() => myCallback(a, b, c, d), (a, b, c, d))
React.useEffect(() => myEffect(a), [a])
React.useMemo(() => myFunction(a, b), (a, b))
```

**Comment by cknitt** [comment]
Ah, sorry, I see you implemented that already, but removed the existing 0-7.

I would suggest to keep them for compatibility with existing projects and maybe deprecate them later on.

**Comment by mununki** [comment]
How does it look? https://github.com/rescript-lang/rescript-react/pull/97/commits/c2e678879b581f057047fb40d83ce07b5bd7b896

If it looks fine, then I'll change the others in same way.

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,9 @@
 #### :boom: Breaking Change
 
 - Requires ReScript 11.0.0-alpha.6 or newer.
+- Deprecated useCallbackN functions in favor of changing `useCallback` signature. (`useCallback: 'f => 'f` -> `useCallback: ('f, 'deps) => 'f`)
+  - Use useCallback instead of useCallbackN. e.g. `useCallback3(f, (a, b, c))` -> `useCallback(f, (a, b, c))`
+  - With this change, it is now possible to pass any value as the second argument `'deps`, but you can still use tuples as before. It provides convenience and flexibility, especially in case of using more than 7 dependencies.
```
Comment (cknitt) [inline CHANGELOG.md:24]:
```suggestion
  - With this change, it is now possible to pass any value as the second argument `'deps`. In case you pass an invalid value, you will get a warning from React at runtime. You should be using one of the following values for the dependency array:
    - 0 dependencies: `[]`
    - 1 argument: `[a]`
    - more than 1 argument: `(a, b, ...)`
```

File: src/React.res [hunk 2]
```diff
@@ -241,31 +241,31 @@ external useMemo6: (@uncurry (unit => 'any), ('a, 'b, 'c, 'd, 'e, 'f)) => 'any =
 external useMemo7: (@uncurry (unit => 'any), ('a, 'b, 'c, 'd, 'e, 'f, 'g)) => 'any = "useMemo"
 
 @module("react")
-external useCallback: 'f => 'f = "useCallback"
+external useCallback: ('f, 'deps) => 'f = "useCallback"
 
-@module("react")
+@module("react") @deprecated("Please use the useCallback instead")
```
Comment (cknitt) [inline src/React.res:246]: "Please use useCallback instead"
or
"Please use the useCallback function instead"

File: CHANGELOG.md (hunk 1 above)
Comment (mununki) [reply CHANGELOG.md]: Thank you!

File: src/React.res (hunk 2 above)
Comment (mununki) [reply src/React.res]: Thank you!
I'll change the others in the same way.

**Comment by mununki** [comment]
For [useMemo](https://react.dev/reference/react/useMemo), the second argument, dependencies, is required, so I think we can fix it the same way as for useCallback. However, [useEffect](https://react.dev/reference/react/useEffect) has optional dependencies, so I don't think we can do the same for it. Do you have any ideas for useEffect? `useEffectN` maybe?

**Comment by cknitt** [comment]
I think `useEffect` without a dependencies array, meaning "perform the effect on every render" is not what you usually want and therefore should not be "the default".

I think we should do it exactly like for `useMemo` and `useCallback` - and provide an additional binding for the case without dependencies. Couldn't think of a good name for that one though. `useEffectOnEveryRender`?

**Comment by cristianoc** [comment]
Why a different function? Can use ()

**Comment by cknitt** [comment]
> Why a different function? Can use [] or ()

Not sure what you mean.

There are three different cases:

1. No dependency array: effect is run on every render
2. Empty dependency array: effect is run on mount
3. Non-empty dependency array: effect is run on mount and when a dependency changes.

**Comment by mununki** [comment]
```res
@module("react")
external useEffect: (@uncurry (unit => option<unit => unit>), 'deps) => unit = "useEffect"

useEffect(_ => {
  Js.log("1")
  None
}, ())

useEffect(_ => {
  Js.log("1")
  None
}, [])
```
```js
import * as React from "react";

React.useEffect((function () {
        console.log("1");
      }), undefined);

React.useEffect((function () {
        console.log("1");
      }), []);
```

**Comment by cknitt** [comment]
> I think we need to make sure it would be fine to pass undefined as a second argument in runtime. I guess it would be fine.

Ah, right I get it now, would be great if it works that way (undefined as second parameter giving the same behavior as invoking the function with a single parameter).

**Comment by mununki** [comment]
Normally, omitting the second argument would be familiar to React developers, and there would be some awkwardness in passing a unit, but I think it's better than the inconvenience of useEffectN. 😄

**Comment by mununki** [comment]
> So if we document this usage, maybe better to pass `None` in the first case which compiles to `undefined`, too?

Yes, same output.

**Comment by mununki** [comment]
Actually, I prefer the unit instead of None. This is because the syntax for optional labeled arguments also utilizes unit.

**Comment by cristianoc** [comment]
> The subtle difference between `()` and `[]` will probably confuse newcomers though, given that we have all these cases then:
> 
> * `()`
> * `[]`
> * `[a]`
> * `(a, b, ...)`
> 
> So if we document this usage, maybe better to pass `None` in the first case which compiles to `undefined`, too?

If it works that's a possibility.
Whatever is chosen, there's no way to "fix" the API which is intrinsically designed to be as confusing as an API can possibly be.

**Comment by cristianoc** [comment]
Screenshot 2023-05-23 at 15 12 13

**Comment by mununki** [comment]
> > One possibility would be to have a descriptive, untagged variant with 2 cases. Might or might not be better.
> 
> Then I'd rather have two separate bindings.
> 
> In any case, we need to keep in mind that `useEffect` without a dependency array is very probably not what you want or should be using anyway. I do not think we are using this in any of our apps and cannot think of a use case for it.
> 
> I think we should just have
> 
> ```rescript
> external useEffect: (@uncurry (unit => option<unit => unit>), 'deps) => unit = "useEffect"
> ```
> 
> as suggested and not even document case 1 (how to pass undefined) at all.

I know this is mostly unused, but I think it's still worth mentioning in the documentation, especially compared to cases like useCallback or useMemo where the second argument is absolutely necessary.

**Comment by mununki** [comment]
> Screenshot 2023-05-23 at 15 12 13

It would be a good use case of untagged variant, but we need one more case for the heterogeneous array or list.

**Comment by cknitt** [comment]
We can also define something like

```rescript
type runEffectOnEveryRender

@val
external runEffectOnEveryRender: runEffectOnEveryRender = "undefined"
```

giving us the following options to document:

* `React.runEffectOnEveryRender`
* `[]`
* `[a]`
* `(a, b, ...)`

**Comment by mununki** [comment]
Since we've already adopted simplicity and flexibility, what about the documentation explains to use `unit` for not having a dependencies?

**Comment by cristianoc** [comment]
I think the tradeoffs are clear. It looks like the case of omitting the second arg cannot be left out without making things worse (People reading the React docs will follow what is recommended).

So the only question is how to represent that case out of the ways considered.
I don't have strong opinions either way.

**Comment by mununki** [comment]
I've tried a rough implementation of heterogeneous array for deps expression:
https://rescript-lang.org/try?version=v10.1.2&code=C4TwDgpgBMA8DkBDAfFAvFJAoLABArgHYBGA9gB4QAmWokUAThAMZRURjpQAiHAFHAD6yAJQAudmBx1okgM5dEDBohCxJyHABsIwNhzlj9YBRgDavMHwCMIgDQ9+AIkRORAXSxA

**Comment by mununki** [comment]
To summarize, we seem to be at 3 options.
* Bind only one function, useEffect, and document the `'dep`.
* Bind two if deps are optional, like useEffect. e.g. `useEffectOnEveryRender`, `useEffect`
* Use an untagged variant to make it a bit more type safe.
All of the above options give us the convenience of not having to use bindings like useEffectN.

**Comment by cristianoc** [comment]
> help finding criteria...

Choosing one of the three options you provided would depend on several criteria from the user's perspective. Here are some important factors to consider:

1. **Ease of Discovery**: How quickly can a new user or developer find and understand how to use these bindings? In terms of ease of discovery, option 2 might be the best. Naming the functions descriptively (like `useEffectOnEveryRender` and `useEffect`) could make them easier to find and understand at a glance.

2. **Error Handling and Debugging**: How easy is it to understand error messages or to debug when something goes wrong? Option 3 might be superior in this regard, as using an untagged variant can lead to more descriptive compile-time errors, helping to find bugs more quickly.

3. **Knowledge Transfer from React**: When considering the ease of transferring knowledge from React, option 1 could be preferred. It matches the original React API more closely, which means that developers already familiar with React's `useEffect` would likely adapt quicker.

4. **Understanding Compiler Errors**: Option 3, using untagged variants, could be advantageous for understanding compiler errors. It can provide more specific compile-time checks, making errors easier to understand and fix.

5. **Learning Curve**: Option 1, binding only one function, `useEffect`, and documenting the dependency, has a relatively low learning curve for those who already know React. But it might confuse beginners due to the optional nature of dependencies. Option 2, binding two separate functions, has a slightly steeper learning curve but can reduce confusion due to explicit naming. Option 3, using an untagged variant, would likely have the steepest learning curve due to introducing the concept of untagged variants.

6. **Consistency**: Is the API consistent with the rest of your codebase? It is often advantageous to choose an option that is more in line with the coding styles and conventions in your existing codebase. 

In conclusion, all these options have their advantages and trade-offs. The ideal choice would depend on the specific needs, preferences, and the level of experience of your team.

**Comment by cristianoc** [comment]
> 1. Bind only one function, useEffect, and document the `'dep`.
> 2. Bind two if deps are optional, like useEffect. e.g. `useEffectOnEveryRender`, `useEffect`
> 3. Use an untagged variant

Is 2 the same as 1 with one extra function that one might or might not use?

**Comment by mununki** [comment]
> > 1. Bind only one function, useEffect, and document the `'dep`.
> > 2. Bind two if deps are optional, like useEffect. e.g. `useEffectOnEveryRender`, `useEffect`
> > 3. Use an untagged variant
> 
> Is 2 the same as 1 with one extra function that one might or might not use?

Indeed. We can rewrite as below:
1. Bind only one function, useEffect, and document the `'dep`.
2. Add bindings without the second argument, e.g. `useEffectOnEveryRender` besides `useEffect` in the first option.
3. Use an untagged variant

**Comment by cknitt** [comment]
I'd say let's prioritize "Knowledge Transfer from React" and "Learning Curve" and go with 1 or 2.

If you prefer 1, I am fine with it as long as we make it clear in the documentation what the difference between `[]` and `()` is and that `()` is not what you usually want.

**Comment by mununki** [comment]
Let's pick 2! In my opinion, option 2 is most balanced 😄 . If so, does the name make sense to you? `useEffectOnEveryRender`?

**Comment by cristianoc** [comment]
Sounds like it's settled

File: CHANGELOG.md [hunk 3]
```diff
@@ -19,6 +19,14 @@
...
 
 - Requires ReScript 11.0.0-alpha.6 or newer.
+- Deprecated use\*N functions in favor of changing the signature of the main hook function.
+  - For example, useEffect instead of useEffectN e.g. `useEffect3(f, (a, b, c))` -> `useEffect(f, (a, b, c))`
+  - The affected hooks include `useEffect`, `useLayoutEffect`, `useCallback`, `useMemo`, `useImperativeHandle`, `useInsertionEffect`
+  - With this change, it is now possible to pass any value as the second argument `'deps`. In case you pass an invalid value, you will get a warning from React at runtime. You should be using one of the following values for the dependency array:
+    - 0 dependencies: `[]`
+    - 1 dependency: `[a]`
+    - more than 1 dependency: `(a, b, ...)`
+    - In case of omitting dependency which means the effect is executed on every render, use \*onEveryRender functions. e.g. `useEffectOnEveryRender`
```
Comment (cknitt) [inline CHANGELOG.md:29]:
```suggestion
- For calling `useEffect`, `useLayoutEffect` etc. *without* a dependency array (meaning that the effect is executed on every render), there are now separate bindings `useEffectOnEveryRender`, `useLayoutEffectOnEveryRender` etc.
```

File: src/RescriptReactRouter.res [hunk 4]
```diff
@@ -190,7 +191,7 @@ let useUrl = (~serverUrl=?, ()) => {
     }
   )
 
-  React.useEffect0(() => {
+  React.useEffectOnEveryRender(() => {
```
Comment (cknitt) [inline src/RescriptReactRouter.res:194]: This should actually be `useEffect(() => ..., [])`

File: src/RescriptReactRouter.res [hunk 5]
```diff
@@ -62,7 +62,8 @@ let arrayToList = a => {
 let pathParse = str =>
   switch str {
   | ""
-  | "/" => list{}
+  | "/" =>
+    list{}
```
Comment (cknitt) [inline src/RescriptReactRouter.res:66]: Why this format change?

**Review by cknitt — CHANGES_REQUESTED** [review]

File: src/RescriptReactRouter.res (hunk 4 above)
Comment (mununki) [reply src/RescriptReactRouter.res:194]: Oh, correct! Good catch.

File: src/RescriptReactRouter.res (hunk 5 above)
Comment (mununki) [reply src/RescriptReactRouter.res:66]: In my editor, vsc, it formats 🤔 The version of extension is v1.18.0

File: CHANGELOG.md (hunk 3 above)
Comment (mununki) [reply CHANGELOG.md]: Thanks!

File: src/RescriptReactRouter.res (hunk 5 above)
Comment (mununki) [reply src/RescriptReactRouter.res:66]: Need to revert it?

Comment (cknitt) [reply src/RescriptReactRouter.res:66]: Ah, right, same for me. Not necessary to revert it, but this formatting is a bit weird, I will report it in the compiler repo.

**Review by cknitt — APPROVED** [review]

File: src/RescriptReactRouter.res (hunk 5 above)
Comment (cknitt) [reply src/RescriptReactRouter.res:66]: Ah no, sorry, it is correct that way, as the line above does not have a `=>`. 🤦

Maybe better to rewrite in one line as `  | "" | "/" => list{}`.
